themes/default/content/blog/iac-recommended-practices-developer-stacks-git-branches/index.md

```diff
@@ -0,0 +1,209 @@
+
+    ```
+    $ pulumi up
+    error: [409] Conflict: Another update is currently in progress.
+    ```
+
+    But on the other, it's a not-so-good thing in that it's a reduction in development velocity: The update that can't run because another update is in progress is an update that will have to wait until sometime later. The more often this happens, the longer it ultimately takes for the team to get the validation it needs to keep moving forward.
+
+So facing these issues, what should the Zephyr team do? The `dev` stack had originally been created for rapid development --- and that worked well, when there was only one developer working on infrastructure. With several, however, things clearly get complicated. Should the team be using a `dev` stack at all? If so, how? And if not, what should they be doing instead?
+
+## Stepping back to reassess
+
+With what we've learned so far, we can return to the questions we raised at the beginning of this post and begin to sketch out a few recommendations for Team Zephyr. Those questions, again, were:
+
+* **How can I best enable multiple developers to collaborate on a Pulumi project?** For Zephyr, an ideal answer to this question would give every member of the team the ability to iterate on both application and infrastructure code concurrently, deploying as often as needed without having to worry about disrupting the work or the velocity of anyone else on the team.
+
+* **How can I use Git and Git branching to support this collaboration?** So far, the feature-branch/pull-request workflow has been serving the Zephyr team well. Should they continue on the same path, merging frequently to `main` and deploying to `dev` and `prod` from there? Or is there some alternative workflow that might serve the team better, either now or at some time in the future?
+
+Let's answer these questions now by taking a look at how the Zephyr team handled them.
+
+### Using developer stacks
+
+We've seen a few ways that having multiple developers share a single stack for ad-hoc deployment can lead to various problems. But we've also seen at least one situation in which being able to run ad-hoc deployments --- without the risk of colliding with existing environments --- can be quite valuable. In fact there are many reasons you might want to do this --- and fortunately Pulumi makes it quite easy to do so.
+
+The recommended practice is to allow developers to create and manage their own Pulumi stacks. In this workflow, developers working on feature branches are able to iterate freely on both application and infrastructure, deploying as often as necessary into a shared cloud environment (typically a sandbox environment, isolated from production) and destroying cloud infrastructure when it's no longer needed.
```

Comment:
i gotta ask, is it not obvious that this would be the solution? in the same way you have branch in git you branch in infra stacks? is this something that people haven't been able to figure out?

Reply:
It's obvious to us, but this stuff gets asked about by users all the time. 

It makes sense if you think about it. Pretty much everyone knows what deployment environments (dev, staging, prod) are in the abstract. But when you pick up a new tool, you don't know how how that tool's concrete features relate to those environments or how people are generally using them to address those environments. Tools like TF don't make this nearly as obvious or easy as Pulumi does, either, so if you're coming to Pulumi from TF, you might not even notice it. So while some people will of course figure it, there's value in making it clear: Here's how it works, here's how people are using it, look how easy it is with Pulumi.

FWIW, this was one of the first things I noticed about Pulumi vs. TF -- i.e., how easy this was vs. TF.